### PR TITLE
Consumer for RSAM amplitude analysis

### PR DESCRIPTION
**Pull request by crockpotveggies**
Wrote a module for RSAM real-time seismic amplitude measurement analysis. The module allows the user to calculate mean, median, min, and max values over a specified time window and optionally forward it to another destination via UDP.

Example from running tests:

```
2020-04-18 03:39:32 TESTING [RSAM] RSAM analysis up and running normally.
2020-04-18 03:39:32 TESTING [RSAM] 2020-01-30 08:26:58 [RSAM] Current RSAM: mean 16280 median 16279 min 16134 max 16456
2020-04-18 03:39:37 TESTING [RSAM] 2020-01-30 08:27:03 [RSAM] Current RSAM: mean 16280 median 16280 min 16145 max 16410
2020-04-18 03:39:43 TESTING [RSAM] 2020-01-30 08:27:08 [RSAM] Current RSAM: mean 16282 median 16284 min 16159 max 16404
2020-04-18 03:39:48 TESTING [RSAM] 2020-01-30 08:27:13 [RSAM] Current RSAM: mean 16283 median 16285 min 16163 max 16418
```

The user can opt to forward RSAM in a lightweight format, JSON, or CSV.

### CODE REVIEW

**Comment by crockpotveggies** [comment]
One thing I'd like to add here is appending the station name to the incoming data. Does anyone know how to extract that information inside the python consumer?

**Comment by iannesbitt** [comment]
> One thing I'd like to add here is appending the station name to the incoming data. Does anyone know how to extract that information inside the python consumer?

Given that our UDP output doesn't include the station name, you should use the user's set station name in `rsudp.raspberryshake.stn`

File: rsudp/c_rsam.py [hunk 1]
```diff
@@ -14,23 +15,28 @@ class RSAM(rs.ConsumerThread):
...
 		"""
 		super().__init__()
 		self.sender = 'RSAM'
 		self.alive = True
 		self.debug = debug
+		self.fwaddr = fwaddr
+		self.fwport = fwport
+		self.sock = False
 		self.interval = interval
 		self.default_ch = 'HZ'
```
Comment (iannesbitt) [inline rsudp/c_rsam.py:44]:
```
self.stn = rs.stn

File: rsudp/c_rsam.py [hunk 2]
```diff
@@ -158,33 +164,49 @@ def _rsam(self):
...
+				self.rsam[3]
 			)
 			printM(msg, self.sender)
 
+	def _forward_rsam(self):
+		"""
+		Send the RSAM analysis via UDP to another destination
+		"""
+		if self.sock:
+			msg = '%s:%s,%s,%s,%s' % (self.cha, self.rsam[0], self.rsam[1], self.rsam[2], self.rsam[3])
```
Comment (iannesbitt) [inline rsudp/c_rsam.py]: To include the station name, change line 195 to the following
```python
 			msg = '%s.%s:%s,%s,%s,%s' % (self.stn, self.cha, self.rsam[0], self.rsam[1], self.rsam[2], self.rsam[3])
```

**Comment by crockpotveggies** [comment]
Hmmm I would hesitate to combine RSAM with Alert because I think that might actually increase confusion. When I see "Alert" I immediately think the functionality should be isolated to alerts only, not an analysis. One possibility here is to rename the RSAM consumer to `Statistics` and open the roadmap to expanding different types of analysis beyond RSAM (such as RSEM and RMS).

I think overall the architecture of rsudp could benefit from composability: being able to chain consumer outputs to each other. I don't necessarily agree with piping messages back to the producer but just my two cents. A future version might look like this:

```
Producer 
    -> StatisticsConsumer 
        -> Alert 
            -> AlertSound
            -> Tweet
            -> Telegram
```

In the meantime we could stick with a separate `Statistics` consumer?

**Comment by iannesbitt** [comment]
That's a really good point. I will do some thinking about reconstructing the flow so that the statistics flow down a separate path from the data. In that case they could even use different instances of the same forward module code. This is a possibility:

```
Producer
    -> Consumer
        -> Statistics
            -> Forward
            -> Alert 
                -> AlertSound
                -> Tweet (still need image paths for social media dissemination)
                -> Telegram (still need image paths for social media dissemination)
        -> Data
            -> Forward
            -> Plot (still need some way of letting the plot know about alerts)
            -> Write
```

This would require some significant changes, so might have to be put into the category of "future", so let's just work on putting out a separate RSAM module for now.

**Comment by crockpotveggies** [comment]
Sounds good, I'll get in some final changes. Can you give me instructions on how to add to the documentation?

File: rsudp/c_rsam.py (hunk 1 above)
Comment (crockpotveggies) [reply rsudp/c_rsam.py:44]: I got a formatting error when attempting this:
```
Traceback (most recent call last):
  File "/home/justin/miniconda3/envs/rsudp/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/home/justin/miniconda3/envs/rsudp/lib/python3.7/site-packages/rsudp/c_rsam.py", line 253, in run
    (self.interval, self.stn, self.cha, self.fwaddr),
TypeError: not all arguments converted during string formatting
```

Is that an object and not a string?

Comment (crockpotveggies) [reply rsudp/c_rsam.py:44]: Nevermind looks like a code issue on my end :)

**Comment by iannesbitt** [comment]
Documentation is built using sphinx. To build the docs, you can do the following:

```bash
pip install sphinx
cd rsudp/rsudp/docsrc
make html
```

Docs will be built to `docsrc/_build/html' for local testing, to build changes to the branch:

```bash
make github
git commit -a -m 'updating docs'
git push
```

With `make github`, docs will be built and copied to the `docs` directory.

**Comment by iannesbitt** [comment]
@crockpotveggies It looks like you added documentation to your docstrings, do you want to build and push the docs before I merge, or would you prefer I merge and build myself?

**Comment by crockpotveggies** [comment]
@iannesbitt go ahead and build the docs on your side if possible, I won't be able to get to it for another few days thanks! :)

**Pull request merged**